sql: hide display of partitions on SHOW CREATE for REGIONAL BY ROW tables

otan opened the pull request:
Resolves #60656 

sql: hide display of partitions on SHOW CREATE for REGIONAL BY ROW tables

This commit hides the partition zone configuration for REGIONAL BY ROW
tables.

This is achieved by the following:
* Introduce a is_multi_region column for crdb_internal.create_statements
* Introduce a builtin
  `crdb_internal.filter_multiregion_fields_from_zone_config_sql` which
  removes multi region zone config fields.
* Filter the partition statements out from the SHOW CREATE TABLE
  delegate function.
* Change REGIONAL BY ROW tests to read from SHOW PARTITIONS so that
  we can still see the zone configurations and partitions.

Release justification: low risk change to new functionality

Release note (sql change): Introduced a new is_multi_region column to
crdb_internal.create_statements, which informs whether the database of
the object is a multi-region database.

Release note (sql change): Introduced a new built-in
`crdb_internal.filter_multiregion_fields_from_zone_config_sql` which
removes multi-region fields from a CONFIGURE ZONE statement.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60761)
<!-- Reviewable:end -->

otan commented:
needs a bit more work

ajstorm reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60761)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-MTv-hyu2MfN-U08QqHe:-MTv-hyu2MfN-U08QqHf:bqj68a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/45081b25e05fcef551264b5f8f4eb2511f533d21/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L43)):*
> ```
> 
> query TT
> SELECT target, raw_config_sql FROM crdb_internal.zones WHERE table_name = 'regional_by_row' ORDER BY target
> ```

How difficult would it be to make this the new `SHOW ZONE CONFIGURATION FOR TABLE`?  Does it accomplish full inheritance?


<!-- Sent from Reviewable.io -->

ajstorm reviewed (commented):
I like this change (though I haven't reviewed all of it yet), but I'm wondering if we can incorporate a change to SHOW ZONE CONFIGURATION FOR TABLE so that we don't have to query the zone configs table manually on each test.  Is that possible?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-:-MTv1Lgd9hwfGqNMsafA:b7lcux)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-MTv-hyu2MfN-U08QqHe:-MU51OuJ1W1rSGx-Vojz:b-m92esp) ([raw file](https://github.com/cockroachdb/cockroach/blob/45081b25e05fcef551264b5f8f4eb2511f533d21/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L43)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

How difficult would it be to make this the new `SHOW ZONE CONFIGURATION FOR TABLE`?  Does it accomplish full inheritance?
</blockquote></details>

not as easy. the logic to ascend to the database descriptor still needs to live concurrently. i'm also worried about the UI of showing a zone config that contains `DATABASE mrdb`, no table zone config, and a lot of partition zone configs. the code atm is also not delegate so it's not a simply SQL query unfortunately.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (changes requested):
Nit: "Introduce _an_ is_multi_region column...".  Also, last bullet point isn't accurate anymore.

> Reviewed 8 of 14 files at r1, 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-:-MUBZotZ3O2ej60ua_mf:b-7z9684)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[docs/generated/sql/functions.md, line 2320 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-MUBfY-m6y4YcCmcJWUN:-MUBfY-m6y4YcCmcJWUO:b-w6k34d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca5caad7397f4dc23527988ae541a24d64f82d48/docs/generated/sql/functions.md#L2320)):*
> ```Markdown
> <tr><td><a name="crdb_internal.filter_multiregion_fields_from_zone_config_sql"></a><code>crdb_internal.filter_multiregion_fields_from_zone_config_sql(val: <a href="string.html">string</a>) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Takes in a CONFIGURE ZONE SQL statement and returns a modified
> SQL statement omitting multi-region related zone configuration fields.
> Returns NULL if no CONFIGURE ZONE statement is required if the
> ```

Nit: See comment below about "i.e.".  Would apply here as well.
___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 43 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-MTv-hyu2MfN-U08QqHe:-MUBaJqU8Qptc9kjhbCm:b-imbr0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/45081b25e05fcef551264b5f8f4eb2511f533d21/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L43)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

not as easy. the logic to ascend to the database descriptor still needs to live concurrently. i'm also worried about the UI of showing a zone config that contains `DATABASE mrdb`, no table zone config, and a lot of partition zone configs. the code atm is also not delegate so it's not a simply SQL query unfortunately.
</blockquote></details>

Looks better!  These would probably read much easier if you selected zone_config first, as the rows below would better line-up.  Also, might be nice to add `colnames` so it's clearer in the output what's being displayed.
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-MUBbTwL2Lt1Qi9tq8XC:-MUBbTwL2Lt1Qi9tq8XD:b9avmwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca5caad7397f4dc23527988ae541a24d64f82d48/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L93)):*
> ```
>                             CONSTRAINT "primary" PRIMARY KEY (pk ASC),
>                             FAMILY fam_0_pk_a_crdb_region (pk, a, crdb_region)
> ) LOCALITY REGIONAL BY ROW
> ```

No need to validate the zone configs here (and for the rest of these CREATE statements)?
___
*[pkg/config/zonepb/zone.go, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-MUBcncy2cQodCMxdPs-:-MUBcncy2cQodCMxdPs0:b4czwxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca5caad7397f4dc23527988ae541a24d64f82d48/pkg/config/zonepb/zone.go#L61)):*
> ```Go
> 
> // MultiRegionZoneConfigFields are the fields on a zone configuration which
> // multi-region objects fully control.
> ```

Nit: "fully control" seems like it might be overkill here, as the user can still override them.  What about "which may be set by the system for multi-region objects"?
___
*[pkg/sql/sem/builtins/builtins.go, line 5003 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-MUBfEkP3mZpt0qTC9bs:-MUBfEkP3mZpt0qTC9bt:b150nsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca5caad7397f4dc23527988ae541a24d64f82d48/pkg/sql/sem/builtins/builtins.go#L5003)):*
> ```Go
> 			`Takes in a CONFIGURE ZONE SQL statement and returns a modified
> SQL statement omitting multi-region related zone configuration fields.
> Returns NULL if no CONFIGURE ZONE statement is required if the
> ```

Nit:  might want to put the last part of this ("...if the multi-region...") in "(i.e. if the ...)"


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60761)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 93 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-MUBbTwL2Lt1Qi9tq8XC:-MUBuolz7Cn2xqrhWjw2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca5caad7397f4dc23527988ae541a24d64f82d48/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L93)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

No need to validate the zone configs here (and for the rest of these CREATE statements)?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-:-MUFzE262TWVi_yK2SUi:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[docs/generated/sql/functions.md, line 2320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60761#-MUFyxqc9c5yVujVvlHV:-MUFyxqc9c5yVujVvlHW:b-elqviw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d3f87c7dd8abef3dc8af4009a69348061738ed6/docs/generated/sql/functions.md#L2320)):*
> ```Markdown
> <tr><td><a name="crdb_internal.filter_multiregion_fields_from_zone_config_sql"></a><code>crdb_internal.filter_multiregion_fields_from_zone_config_sql(val: <a href="string.html">string</a>) &rarr; <a href="string.html">string</a></code></td><td><span class="funcdesc"><p>Takes in a CONFIGURE ZONE SQL statement and returns a modified
> SQL statement omitting multi-region related zone configuration fields.
> If the CONFIGURE ZONE statement can be inferred by the database or table,
> ```

Nit: "can be inferred by the database's or table's zone configuration, this will return NULL"?


<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm

otan commented:
bors r-

otan committed:
sql: hide display of partitions on SHOW CREATE for REGIONAL BY ROW tables

This commit hides the partition zone configuration for REGIONAL BY ROW
tables.

This is achieved by the following:
* Introduce a is_multi_region column for crdb_internal.create_statements
* Introduce a builtin
  `crdb_internal.filter_multiregion_fields_from_zone_config_sql` which
  removes multi region zone config fields.
* Filter the partition statements out from the SHOW CREATE TABLE
  delegate function.
* Change REGIONAL BY ROW tests to read from SHOW PARTITIONS so that
  we can still see the zone configurations and partitions.

Release justification: low risk change to new functionality

Release note (sql change): Introduced a new is_multi_region column to
crdb_internal.create_statements, which informs whether the database of
the object is a multi-region database.

Release note (sql change): Introduced a new built-in
`crdb_internal.filter_multiregion_fields_from_zone_config_sql` which
removes multi-region fields from a CONFIGURE ZONE statement.

otan commented:
bors r=ajstorm

otan commented:
bors r=ajstorm
